### PR TITLE
Feature/blockrobot tactic

### LINKED ISSUE #859: BlockRobot
Gets a robot to block from the play.
Goes to the position of the robot
Stands in between that robot and the ball

Reset: Robot no longer on line between enemy robot and target

### PR DESCRIPTION
**Pull request by RobotJente**
Seems to do what it's supposed to. Hard to test since robot_hub won't send my commands, so not sure about whether the shouldTacticReset works properly, but I also can't see any mistakes.

Edit: I thought there was a mistake somewhere on development causing robothub to freak out but apparently it is my code.

Closes #859

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -175,6 +175,8 @@ set(NEW_SKILLS_SOURCES
 set(NEW_TACTICS_SOURCES
         ${PROJECT_SOURCE_DIR}/src/stp/Tactic.cpp
         ${PROJECT_SOURCE_DIR}/src/stp/new_tactics/TestTactic.cpp
+        ${PROJECT_SOURCE_DIR}/src/stp/new_tactics/BlockRobot.cpp
+
```
Comment (timovdk) [inline CMakeLists.txt]:
```suggestion
```

File: include/roboteam_ai/stp/StpInfo.h [hunk 2]
```diff
@@ -145,6 +153,13 @@ struct StpInfo {
      * Speed of the dribbler in %
      */
     int dribblerSpeed = 0;
+
+    /**
+     * When blocking off a position, the robot is on line between a targetposition to block, and the enemy robot.
+     * Used to decide how close this robot should be to enemy robot
+     */
+    BlockDistance blockDistance;
+
```
Comment (timovdk) [inline include/roboteam_ai/stp/StpInfo.h]:
```suggestion
```

File: include/roboteam_ai/stp/new_tactics/BlockRobot.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+         * @param targetLocation the location that you want to block off from the robot. For example, the ball position,
+         * or our goal
+         * @return the desired position to block the target.
+         */
+        Vector2 calculateDesiredRobotPosition(BlockDistance blockDistance, const world_new::view::RobotView enemy,
+                                              Vector2 targetLocation);
+
+        virtual bool isTacticFailing(const StpInfo &info) noexcept;
+        virtual bool shouldTacticReset(const StpInfo &info) noexcept;
+
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_tactics/BlockRobot.h]:
```suggestion
```

File: src/stp/new_tactics/BlockRobot.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace rtt::ai::stp::tactic {
+
+    BlockRobot::BlockRobot() {
+        skills = rtt::collections::state_machine<Skill, Status, StpInfo>{skill::GoToPos(), skill::Rotate()};
+        skills.initialize();
+    }
+
+    void BlockRobot::onInitialize() noexcept {
+
```
Comment (timovdk) [inline src/stp/new_tactics/BlockRobot.cpp]:
```suggestion
```

File: src/stp/new_tactics/BlockRobot.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    void BlockRobot::onUpdate(Status const &status) noexcept {
+
```
Comment (timovdk) [inline src/stp/new_tactics/BlockRobot.cpp]:
```suggestion
```

File: src/stp/new_tactics/BlockRobot.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        auto desiredRobotPosition = calculateDesiredRobotPosition(info.getBlockDistance(), info.getEnemyRobot().value(), info.getTargetPos().second);
+        skillStpInfo.setTargetPos(std::make_pair(TargetType::MOVETARGET, desiredRobotPosition));
+
+        return skillStpInfo;
+    }
+
+    double BlockRobot::calculateAngle(const world_new::view::RobotView enemy, Vector2 targetLocation) {
+        Vector2 lineEnemyToTarget = targetLocation - enemy->getPos();
+        return lineEnemyToTarget.angle();
+
```
Comment (timovdk) [inline src/stp/new_tactics/BlockRobot.cpp]:
```suggestion
```

File: include/roboteam_ai/stp/new_tactics/BlockRobot.h [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+namespace rtt::ai::stp::tactic {
+    class BlockRobot : public Tactic {
+    public:
+        BlockRobot();
+        virtual bool isEndTactic() noexcept;
+
+    private:
+        // TODO: make this a sensible margin
+        const double errorMargin = 4*Constants::ROBOT_RADIUS();
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_tactics/BlockRobot.h]: You mean to figure this margin out during testing?

File: src/utilities/Constants.cpp [hunk 8]
```diff
@@ -108,7 +108,7 @@ double Constants::MAX_RECEIVE_TIME() { return 1.0; }
 
 double Constants::BALL_STILL_VEL() { return 0.1; }
 
-double Constants::GOTOPOS_ERROR_MARGIN() { return 0.03; }
+double Constants::GOTOPOS_ERROR_MARGIN() { return 0.3; }
```
Comment (RobotJente) [inline src/utilities/Constants.cpp]:
```suggestion
double Constants::GOTOPOS_ERROR_MARGIN() { return 0.03; }
```

File: src/stp/Tactic.cpp [hunk 9]
```diff
@@ -27,6 +27,9 @@ Status Tactic::update(StpInfo const &info) noexcept {
 
     // the tactic will not be reset if it's the first skill; it will be reset as well if the
     // state machine is finished
+    auto temp = skills.finished();
+    auto c = skills.current_num();
+    auto d = shouldTacticReset(info);
```
Comment (RobotJente) [inline src/stp/Tactic.cpp]:
```suggestion
```

File: src/stp/new_plays/TestPlay.cpp [hunk 10]
```diff
@@ -51,6 +51,8 @@ void TestPlay::assignRoles() noexcept {
             stpInfos[roleName].setField(*world->getField());
 
             // TODO calculate additional info
+
```
Comment (RobotJente) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
```

File: src/stp/new_plays/TestPlay.cpp [hunk 11]
```diff
@@ -51,6 +51,7 @@ void TestPlay::assignRoles() noexcept {
...
```
Comment (RobotJente) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
```

**Pull request merged**